server/trpc/modules/redemption/redemption.router.ts

```diff
@@ -0,0 +1,94 @@
+import { TRPCError } from '@trpc/server'
+import { defaultRedemptionSelect } from '../redemption/redemption.select'
+import { protectedProcedure, router } from '~/server/trpc/trpc'
+import { redeemAdminClaimInput, redeemClaimInput } from '~/shared/voucher'
+
+export const redeemRouter = router({
+
+  list: protectedProcedure
+    .query(async ({ ctx }) => {
+      try {
+        return await ctx.prisma.redemption.findMany({
+          where: {
+            userId: ctx.session.user.id,
+          },
+          select: defaultRedemptionSelect,
+        })
+      }
+      catch (err) {
+        ctx.logger.error({ msg: 'failed to list redeemed vouchers', err })
+        throw new TRPCError({
+          code: 'INTERNAL_SERVER_ERROR',
+          message: 'Failed to list redeemed vouchers',
+        })
+      }
+    }),
+
+  listVoucherRequireRedeem: protectedProcedure
+    .query(async ({ ctx }) => {
+      try {
+        return await ctx.prisma.redemption.findMany({
+          where: {
+            claimed: false,
+          },
+          select: defaultRedemptionSelect,
+        })
+      }
+      catch (err) {
+        ctx.logger.error({ msg: 'failed to list redeemed vouchers', err })
+        throw new TRPCError({
+          code: 'INTERNAL_SERVER_ERROR',
+          message: 'Failed to list redeemed vouchers',
+        })
+      }
+    }),
+
+  claim: protectedProcedure
+    .input(redeemClaimInput)
+    .mutation(async ({ ctx, input }) => {
+      try {
+        return await ctx.prisma.redemption.update({
+          where: {
+            id: input.redemptionId,
+            userId: ctx.session.user.id,
+          },
+          data: {
+            claimed: true,
+          },
+        })
+      }
+      catch (err) {
+        ctx.logger.error({ msg: 'failed to claim voucher', err })
+        throw new TRPCError({
+          code: 'INTERNAL_SERVER_ERROR',
+          message: 'Failed to claim voucher',
+        })
+      }
+    },
+    ),
+
+  adminClaim: protectedProcedure
```

Comment:
Why need admin claim?

Reply:
This is to allow admin to essentially claim on the admin end for the vouchers where as the normal claim is for the users to claim on their end

Reply:
In this case then it is probably better to combine into single claim function and assertInstitutionRole to see if user is admin

Reply:
The idea is for users to be able to claim their redeemed voucher right? 

Say I redeem a LIHO voucher. I will need to go to an admin to get the voucher code. Then the admin will mark the voucher as claimed right?

What is the use case for members claiming their own voucher?

Reply:
If you have seen how Shopee system can also work where the user is able to claim on their end at the same time on the shopee administrator end, they are also able to mark the voucher as claimed

Reply:
I don’t think it applies since we have one claimed field only